src/promptflow-tools/promptflow/tools/open_source_llm.py

```diff
@@ -0,0 +1,466 @@
+            """
+            raise KeyError(message)
+
+
+class LlamaContentFormatter(ContentFormatterBase):
+    """Content formatter for LLaMa"""
+
+    def __init__(self, api: API, chat_history: Optional[str] = ""):
+        super().__init__()
+        self.api = api
+        self.chat_history = chat_history
+
+    @staticmethod
+    def validate_role(role: str) -> None:
+        valid_roles = {"system", "user", "assistant"}
+        if role not in valid_roles:
+            valid_roles_str = ",".join([f"'{role}:\\n'" for role in valid_roles])
+            error_message = f"""The Chat API requires a specific format for prompt definition, and the prompt should
+            include separate lines as role delimiters: {valid_roles_str}. Current parsed role '{role}' does not
+            meet the requirement. If you intend to use the Completion API, please select the appropriate API type
+            and deployment name. If you do intend to use the Chat API, please refer to the guideline at
+            https://aka.ms/pfdoc/chat-prompt or view the samples in our gallery that contain 'Chat' in the name."""
+            raise Exception(error_message)
+
+    @staticmethod
+    def parse_chat(chat_str: str) -> List[Dict[str, str]]:
+        # LLaMa only supports below roles.
+        separator = r"(?i)\n*(system|user|assistant)\s*:\s*\n"
+        chunks = re.split(separator, chat_str)
+        chat_list = []
+        for chunk in chunks:
+            last_message = chat_list[-1] if len(chat_list) > 0 else None
+            if last_message and "role" in last_message and "content" not in last_message:
+                last_message["content"] = chunk
+            else:
+                if chunk.strip() == "":
+                    continue
+                role = chunk.strip().lower()
+                # appends consecutive user messages together to support workaround for system role
+                if last_message and last_message.get("role", "") == "user" and role not in {"user", "assistant"}:
+                    last_message["content"] += "\\n" + chunk
+                    continue
+                # Check if prompt follows chat api message format and has valid role.
+                LlamaContentFormatter.validate_role(role)
+                if last_message and last_message.get("role", "") == "user" and role == "user":
+                    continue
+                # system role not officially supported, so we hack a solution by making system role the user role
+                new_message = {"role": role} if role != "system" else {"role": "user"}
+                chat_list.append(new_message)
+
+        return chat_list
+
+    def format_request_payload(self, prompt: str, model_kwargs: Dict) -> bytes:
+        """Formats the request according the the chosen api"""
+        if "do_sample" not in model_kwargs:
+            model_kwargs["do_sample"] = True
+
+        if self.api == API.CHAT:
+            request_payload = render_jinja_template(
```

Comment:
Why the other implementation directly construct a dictionary but we render jinja here?